### ISSUE TITLE
DnsNameResolver don't use provided executorPool

### ISSUE THREAD

**Issue by Gordiychuk** [grpc/grpc-java#3703]
### What version of gRPC are you using?
v1.7.0

### What JVM are you using (`java -version`)?
java version "1.8.0_112"
Java(TM) SE Runtime Environment (build 1.8.0_112-b15)
Java HotSpot(TM) 64-Bit Server VM (build 25.112-b15, mixed mode)

### What did you do?
```java
        ExecutorService customExecutor = Executors.newFixedThreadPool(2, new DefaultThreadFactory("grpc-client-test"));
        EventLoopGroup customEventLoop = new EpollEventLoopGroup(1);

        ManagedChannel channel = NettyChannelBuilder.forAddress("localhost", 8080)
                .eventLoopGroup(customEventLoop)
                .executor(customExecutor)
                .build();

        MyServiceGrpc.MyServiceStub stub = MyServiceGrpc.newStub(channel);
        // call any method on stub
        // ...
```

### What did you expect to see?
Will not be created additional thread for rpc call expect configured before.

### What did you see instead?
In my case grpc still use default CachedThreadPool and creates at least 32 thread. 
```
solomon-alert_1  | "grpc-default-executor-31" #124 daemon prio=5 os_prio=0 tid=0x00007f1762851800 nid=0xb3 waiting on condition [0x00007f16ec3cf000]
solomon-alert_1  |    java.lang.Thread.State: TIMED_WAITING (parking)
solomon-alert_1  |      at sun.misc.Unsafe.park(Native Method)
solomon-alert_1  |      - parking to wait for  <0x00000005923d3148> (a java.util.concurrent.SynchronousQueue$TransferStack)
solomon-alert_1  |      at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
solomon-alert_1  |      at java.util.concurrent.SynchronousQueue$TransferStack.awaitFulfill(SynchronousQueue.java:460)
solomon-alert_1  |      at java.util.concurrent.SynchronousQueue$TransferStack.transfer(SynchronousQueue.java:362)
solomon-alert_1  |      at java.util.concurrent.SynchronousQueue.poll(SynchronousQueue.java:941)
solomon-alert_1  |      at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1066)
solomon-alert_1  |      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
solomon-alert_1  |      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
solomon-alert_1  |      at java.lang.Thread.run(Thread.java:745)
```

It's occurs because DnsNameResolverProvider creates DnsNameResolver with default shared executor, instead of use provided for ManagedChannel: https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/DnsNameResolverProvider.java#L51

**Comment by ejona86** [comment]
This was on purpose, because DNS resolution is blocking. If the application is async, then DNS resolution blocking threads could be quite surprising. I think C creates a thread here as well, although they limit to one DNS resolution at a time; we could do that as well by using a SerializingExecutor.

@Gordiychuk, what problem is this causing. Is it because you're on Android?

CC @zhangkun83.

**Comment by ejona86** [comment]
> In my case grpc still use default CachedThreadPool and creates at least 32 thread.

Oh, nm. I guess that's the problem. Yeah, using SerializingExecutor would bring that down to 1. Would that resolve your concern?

**Comment by Gordiychuk** [comment]
> Yeah, using SerializingExecutor would bring that down to 1. Would that resolve your concern?

One thread will be enough? I think ability to customize thread pool for DnsNameResolver will be better.

**Comment by ejona86** [comment]
> One thread will be enough?

Much of the time. Although yeah, it could be a problem in some workloads. I guess an option is simply to limit it to, say 4. That wouldn't be too bad in many workloads (those that only need 1) while still allowing more. Dunno.

I talked with C core, and they no longer use a single thread, but the change was not intentional in relation to this discussion.

Maybe this could be mixed with #3644, where you construct a DnsNameResolver with a custom executor and register it. Unfortunately that would mean exposing DnsNameResolver as part of our API.

**Comment by bartlomiej-dawidow** [comment]
This is related to #1400

**Comment by groakley** [comment]
We are also running into this issue. Our Android environment strictly limits the number of threads that are created, and we prohibit libraries that spin up their own ThreadPools. We have a designated ThreadPool for blocking work that we would prefer to pass to grpc for this purpose when we create Channels.

As proposed by @Gordiychuk, we would like to be able to optionally override the default executor with an additional method on __ChannelBuilder, e.g. `CronetChannelBuilder#blockingExecutor(ExecutorService)`

**Comment by ejona86** [comment]
Specifying a `blockingExecutor()` may be the best approach (although we may still want to clean up the 32 thread explosion that was seen originally). We may need to use that for CallCredentials as well, since I think we are maybe incorrectly doing blocking calls on the user's thread in that case. We'd need to figure out an API to pass the executor to the NameResolver.

@zhangkun83, thoughts?

**Comment by zhangkun83** [comment]
SGTM. It also sounds more appropriate for `CallCredentials` to use the blocking executor rather than the app executor.

**Comment by groakley** [comment]
@ejona86 Thank you for the clarification. That was my mistake: we are only seeing this issue for Android clients that use the OkHttp transport. My example should have used the `ManagedChannelBuilder` instead.

**Comment by ejona86** [comment]
@groakley, ah, thanks. Makes sense. It's easy to get the two mixed up when using both. So this remains an issue for OkHttp and Netty.

**Comment by ericgribkoff** [comment]
@groakley I spoke with @ejona86 and this is definitely open for you to send a PR if you'd like to pursue this. Something like a new `blockingExecutor()` API on the builder that gets plumbed via the [brand new](https://github.com/grpc/grpc-java/commit/e875a8c6a39f1ec8e2fba581e87397295f92e144)  [`NameResolver.Helper`](https://github.com/grpc/grpc-java/blob/e875a8c6a39f1ec8e2fba581e87397295f92e144/core/src/main/java/io/grpc/NameResolver.java#L213) should work.

**Comment by dapengzhang0** [comment]
@groakley, @ejona86 is this issue resolved by #6238, and can be closed?

**Comment by groakley** [comment]
@dapengzhang0 the `blockingExecutor` (possibly be renamed `offloadExecutor`, see #6279) provided to the AbstractChannelBuilder will now be used by the DnsNameResolver. Threads will not be created by a default grpc threadpool if the option is set. 

#6238 did not address the fact that 32 threads were observed. That may make sense to file as a separate issue.

**Comment by ejona86** [comment]
Let's close and if someone has problems with number of concurrent DNS requests and is using the default executor we can use a new issue.
